### PR TITLE
Setup use of allow raw append/prepend

### PR DESCRIPTION
**Pull request by cwood**
With this change we should be able to do conditionals in the begining or end of the pool. I need this for setting up razor.

```
if exists user-class and option user-class = "iPXE" {
    filename "bootstrap.ipxe";
  } else {
    filename "undionly.kpxe";
  }
```

### CODE REVIEW

**Comment by brandonweeks** [comment]
LGTM. Tests might be useful though.

File: templates/dhcpd.pool.erb [hunk 1]
```diff
@@ -2,6 +2,9 @@
 # <%= @name %>
 #################################
 subnet <%= @network %> netmask <%= @mask %> {
+<% if @raw_prepend %>
+   <%= @raw_prepend %>
```
Comment (ekohl) [inline templates/dhcpd.pool.erb]: nitpick: there are 3 spaces while the rest has 2 spaces.

File: templates/dhcpd.pool.erb [hunk 2]
```diff
@@ -65,5 +68,7 @@ subnet <%= @network %> netmask <%= @mask %> {
 <% if @pxeserver -%>
   next-server <%= @pxeserver %>;
 <% end -%>
+<% if @raw_append %>
+   <%= @raw_append %>
```
Comment (ekohl) [inline templates/dhcpd.pool.erb]: nitpick: there are 3 spaces while the rest has 2 spaces.

**Comment by mmoll** [comment]
@lzap as this seems to touch your UEFI work, could you have a look here?
Personlly I'm generally against the possibility to just inject raw text into puppet modules, but maybe we don't have a better option here.

**Comment by ekohl** [comment]
@mmoll I agree that this isn't generally wanted, but in this case I think it's the only solution. You can use concat to hide the problem, but that's doesn't make it any different.

**Comment by brandonweeks** [comment]
@mmoll we've had some pretty crazy stuff we've had to insert for various networking gear. I've tried to extend the module for most of what we need but I see some value in allowing overrides.

**Comment by lzap** [comment]
I cannot review puppet implementation, but this is going to be added soon once we have UEFI implemented:

        if option arch = 00:06 {
            filename "grub2/bootia32.efi";
        } else if option arch = 00:07 {
            filename "grub2/bootx64.efi";
        } else {
            filename "pxelinux.0";
        }

Ideally the requested block from this PR is hardcoded and enabled by special variable. We should modularize filename as well, ideally for all supported arches for the future: https://tools.ietf.org/html/rfc4578#section-2.1 (0, 6, 7, 9).

See http://projects.theforeman.org/projects/foreman/wiki/PXE_Booting_UEFI for more details.

**Comment by lzap** [comment]
To elaborate on that, if you could add these options and make one big modular if statement according to these to something like:

    if exists user-class and option user-class = "iPXE" {
        filename "<%= @filename_ipxe =>";
    } else if option arch = 00:06 {
        filename "<%= @filename_efi32 =>";
    } else if option arch = 00:07 {
        filename "<%= @filename_efibc =>";
    } else {
        filename "<%= @filename =>";
    }

Default values should be all pxelinux.0 for now, until we change our manifest to also deploy other firmware there.

**Comment by cwood** [comment]
Updated with tests and fix for the space issue. Hopefully this is good to be merged.

File: templates/dhcpd.pool.erb [hunk 3]
```diff
@@ -2,6 +2,9 @@
 # <%= @name %>
 #################################
 subnet <%= @network %> netmask <%= @mask %> {
+<% if @raw_prepend %>
```
Comment (ekohl) [inline templates/dhcpd.pool.erb]: Won't this need a ```-%>``` to prevent an empty line in every config?

File: templates/dhcpd.pool.erb [hunk 4]
```diff
@@ -65,5 +68,7 @@ subnet <%= @network %> netmask <%= @mask %> {
 <% if @pxeserver -%>
   next-server <%= @pxeserver %>;
 <% end -%>
+<% if @raw_append %>
```
Comment (ekohl) [inline templates/dhcpd.pool.erb]: Won't this need a ```-%>``` to prevent an empty line in every config?

**Comment by ekohl** [comment]
I think we can merge this. If we need more complex configs, we can move to a concat configuration in an API-compatible way.

File: templates/dhcpd.pool.erb (hunk 4 above)
Comment (cwood) [inline templates/dhcpd.pool.erb]: shouldnt. At least based on my tested output it doesnt add a new line.

**Comment by ekohl** [comment]
Thanks @cwood!

**Comment by lzap** [comment]
On thanks one more deatil, these raw strings must not contain curly braces, otherwise this breaks our current parser. Dmitri is writing new one, but unless this is done... take care!

**Pull request closed without merging**